### PR TITLE
DynISF unified (wip)

### PR DESCRIPTION
**Pull request by justmara**
Idea:

Current Dynamic ISF plugin does not bring anything new to oref1 itself. It only changes the way ISF is calculated. Commonly used ISF modes:
1. static profile-based - simple to use, for average users
2. TDD-based - proof of concept, seems to be working fine.
3. BG based (like in AutoISF) - many people simply want AAPS to act stronger and adjust CSF with BG rise, because they noted that higher BG levels require more insulin for every BG point to deal with.

In details.

TDD-based Dynamic ISF in reality consists of two steps:
1. calculate `base` (normalTarget) ISF from TDD
2. calculate `current` ISF as shift from base one to current BG

Step two seems totally like `BG based ISF` itself.
So I've decided to incorporate all the ISF calculation in one place with following options available:

1. Use Dynamic ISF (on/off)
2. Use TDD for ISF (on/off)
3.    TDD ISF adjustment factor (% scale of dynisf)
4. Scaling velocity (how much BG level affects ISF)
5. Adjust sens (ISF change affects ratio too)
6. Follow the profile changes for ISF calculation (currently DynISF ignores profile changes and you don't have any power over it, being forced to wait till TDD stabilizes).
7. BG cap (level at which ISF scaling applies diminishing returns)

image

At the same time all the JS ISF calculation code is moved to single JS-string injected from `DetermineBasalAdapterJS` so it can be reused in any other oref1-based modification like EN/Boost/etc, so that mods no more require to copy-paste ISF calculations and can focus of their own logic.

### CODE REVIEW

**Comment by koelewij** [comment]
I always was quite happy with the behavior to higher BG’s that DynISF has. I have used it in dev for about 7 months. I quit it because my TDD started fluctuating due to illness and after that I ended up with a roller coaster effect. Tho option to use profile ISF and having a stronger ISF at higher BG’s sounds like a very good option. I see the ISF calculated based on TDD as a nice reference, but constantly updating ISF using TDD is a bit too much. 
I vote positive for this change.

File: plugins/sync/src/main/java/info/nightscout/plugins/sync/nsclientV3/workers/LoadBgWorker.kt [hunk 1]
```diff
@@ -64,7 +64,7 @@ class LoadBgWorker(
                         // Objective0
                         sp.putBoolean(info.nightscout.core.utils.R.string.key_objectives_bg_is_available_in_ns, true)
                         // Schedule processing of fetched data and continue of loading
-                        continueLoading = !(sgvs.size != NSClientV3Plugin.RECORDS_TO_LOAD || response.code == 304)
+                        continueLoading = !(sgvs.size <= NSClientV3Plugin.RECORDS_TO_LOAD*0.95 || response.code == 304)
```
Comment (justmara) [inline plugins/sync/src/main/java/info/nightscout/plugins/sync/nsclientV3/workers/LoadBgWorker.kt:67]: sorry, commited this by an accident. its a crude temporal fix for [this](https://github.com/nightscout/AndroidAPS/issues/2763)

**Comment by jbr7rr** [comment]
Do I read this PR correctly and have you moved the settings to the Sensitivity (autosens) settings? 

While I like the idea of moving it to a separate place so it can be used by other plugins in the future as well, not sure if from a UX perspective the sensitivity plugin is the right place to put these settings it can further complicate the understanding of the system for the user.

Also can you outline which dynISF functionalitiy is currently implemented and what modes are you adding ?

**Comment by justmara** [comment]
Yes, I moved it to sensitivity section, right near autosens settings. Because currently `sensitivity` is taken by autosens in AAPS and it would add a lot of confusion if another `sensitivity` plugin is added. But surely its only UX thing and can be moved to any better place.

Current (dev-i state) DynISF is TDD+BG scaling with TDD adjustment factor and `adjust sens` (when recent tdd change affects `autosens.ratio`).

I've added:

1. Non-TDD (profile one) with BG level scaling mode.
2. BG cap (taken from Boost) to control BG/TDD scaling on higher BG levels
3. Velocity - to control whole impact of BG on ISF scaling (both for TDD and Non-TDD modes)
4. `Obey profile` - so profile % affects both TDD and Non-TDD calculated ISF

Extracted ISF calculation function from `determine-basal.js` so it can be altered from kotlin code.

**Comment by faldor20** [comment]
This sounds really good to me and addresses many of the issues we have switching to dynisf. 
I think the current implementation makes a lot of assumptions about your situation and this brings the customisation needed to make it work much better for people outside the "norm", which is an admirable goal.

**Comment by szantos** [comment]
+1 for it to get merged!

More and more demand (issues) rise because of various controversies regarding DynISF.

Every option can be useful, but let me point out some aspects:

There were more users in the (early) DynISF Discord community, who could not profit of the TDD dependency of DynISF. Problems like site leakage could also need to disconnect TDD from ISF.

Obey/follow profile change % introduces a new way for the user to affect the ISF, which is currently only manually through the DynISF adjustment factor possible. This opens the way towards automation, too.

**Comment by justmara** [comment]
updated to reflect latest dev changes.
moved DyniSF settings back to Dynamic ISF plugin settings but still within its own section.
unified a bit openapssmsplugin/determinebasal inheritance to avoid copy-paste

**Pull request closed without merging**